[FEAT] 모듈 분리

waitAmhee opened the pull request:
## #️⃣연관된 이슈

> #86 

## 📝작업 내용

> 웹소켓용 모듈 분리 

### 스크린샷 (선택)

## 💬리뷰 요구사항(선택)

> 배포 테스트 진행

waitAmhee committed:
feat : 그라파나용 매트릭 -> 처리시간 p50, p95, p99 추가

waitAmhee committed:
feat : 지연 시간 측정을 위해 stomp 헤더에 타임스탬프 추가

waitAmhee committed:
feat : 동시 접속자 수 매트릭 추가

waitAmhee committed:
feat : 메시지 유실 측정을 위한 보낸 메시지 수 매트릭 추가

waitAmhee committed:
refactor : WebSocketMetrics와 WebSocketEventListener는 구체적인 기술/라이브러리 사용하기 때문에 infra 레벨의 관심사로 보여서 위치 이동 + 순환참조 문제를 해결하기 위해 @Lazy 속성 사용

waitAmhee committed:
fix : Prometheus가 메트릭을 다 가져가고 연결을 닫았는데, 서버는 아직 write 중이라서 Broken pipe 발생 -> 무시 처리( 로그 정리를 위한 코드)

waitAmhee committed:
refactor: exceptionHandler에서 AsyncRequestNotUsableException 핸들링 안하는 방향으로 수정

waitAmhee committed:
fix : 환율 api를 사용해 조회하는 방식에서 레디스에 저장한 값을 읽어오는 방식으로 수정

waitAmhee committed:
refactor: 로그 정리

waitAmhee committed:
fix: 헤더에 sentTimestamp가 담기지 않아서 setNativeHeader로 대체

waitAmhee committed:
feat: actuator/metrics 엔드포인트 추가

happyAndy0412 committed:
ehm: jvm에 할당하는 메모리 512Mb로 수정

happyAndy0412 committed:
ehm: jvm에 할당하는 초기 힙 메모리도 512Mb로 수정

happyAndy0412 committed:
ehm: 1. 힙메모리 설정 비교위해 기본으로 복구 2. WebSocketMessageBrokerStats를 사용하여 웹소켓 커스텀 메트릭 설정 3. 웹소켓 브로커 로그 발생 간격 30분 -> 30초로 수정

happyAndy0412 committed:
ehm: 메트릭 명명 규칙 위배 사항 수정

happyAndy0412 committed:
ehm: 브로커 로깅 주기 적용안돼서 직접 설정 파일 생성

happyAndy0412 committed:
ehm: 메트릭 명명규칙 위반(같은 이름은 같은 타입이어야 함) 재수정

waitAmhee committed:
feat: 모듈 분리를 위한 설정 파일 추가 및 수정(websocket 모듈용 application.yml, build.gradle, dockerfile, deploy.yml settings.gradle 수정 등)

waitAmhee committed:
refactor: 기존의 코드 infra 모듈에서 websocket 모듈로 이동

# Conflicts:
#	websocket/src/main/java/com/fintory/websocket/monitoring/config/StompStatsMetricsConfiguration.java
#	websocket/src/main/java/com/fintory/websocket/monitoring/config/WebSocketStatsConfig.java

waitAmhee committed:
feat : LiveStockPriceWebSocketService 클래스 분리

waitAmhee committed:
refactor: 필요없는 코드 삭제 및 수정

waitAmhee committed:
refactor: 기존에는 같은 서버 내에서만 이벤트를 발행/구독 -> Redis Pub/Sub을 통해 서버 간 통신이 가능하도록 구조를 수정

waitAmhee committed:
refactor: 서버-클라, 증권사-서버 패키지 분리

waitAmhee committed:
refactor: 로그정리 및 서버-클라, 증권사-서버 패키지 분리 이후 -> 자동 패키지 리팩토링된 클래스

waitAmhee committed:
fix: predestroy 실행 전에 redis가 닫혀버려서 redisTemplate 접근 중 에러 발생 -> Redis 관련 cleanup 단계 스킵하도록 처리

waitAmhee committed:
feature: websocket 용 build.gradle파일 추가

waitAmhee committed:
refactor: DockerFile 이름 변경

waitAmhee committed:
fix: ecr repository를 fintory로 통일, 태그로 구분

waitAmhee committed:
fix: ecr repository를 fintory-child로 통일

waitAmhee committed:
fix: application.yml -> 프로필명 통일

waitAmhee committed:
refactor: 필요없는 코드 삭제

waitAmhee committed:
fix :  EC2 간 Bean 의존성 분리 (Service/Repository 직접 참조 제거)

waitAmhee committed:
fix : child 모듈에서 websocket 모듈 삭제

waitAmhee committed:
fix : appcliation.yml 변수 활용 방식 변경

waitAmhee committed:
fix : 들여쓰기 에러 수정

waitAmhee committed:
fix : 빈 의존성 수정

waitAmhee committed:
fix : WebSocket 종료 시 Redis 연결 에러 해결

waitAmhee committed:
fix : @EnableJpaAuditing 추가

waitAmhee committed:
fix : actuator 의존성 추가 + 토큰 발급 주기 23시간 -> 8시간으로 변경

waitAmhee committed:
fix : SecurityConfig 추가 -> 그라파나 대시보드 추가를 위해서.

waitAmhee committed:
fix : dependson 제거

waitAmhee committed:
fix : redis replica 설정 해제

waitAmhee committed:
fix : redis replica 설정 추가(테스트)

happyAndy0412 commented on app-child/src/main/java/com/fintory/child/domain/stock/controller/common/CommonStockControllerImpl.java:
```diff
@@ -20,22 +21,27 @@
     @GetMapping("/opened-market")
     public ResponseEntity<ApiResponse<MarketStatusResponse>> getMarketStatus(){
-        return ResponseEntity.ok(ApiResponse.ok(websocketService.getMarketStatus()));
+        String url = "http://"+websocketServerUrl+":8080" + "/api/websocket/market/status";
```
이 로직을 websocket 모듈 내에 있는 controller에 정의하지 않으신 이유가 있으신가요

waitAmhee replied:
처음에는 웹소켓 보다는 일반 장시간을 알려주는 기능이니 주 모듈의 common 컨트롤러에서 처리하는 것이 의미적 일관성이 있다고 판단했는데, 생각해보니 불필요한 HTTP 호출이 발생해서 Websocket 모듈로 옮기는 것이 더 효율적이네요!  수정하도록 하겠습니다! 👍

waitAmhee replied:
아 근데 혹시 자주 호출되는 기능이면 현재 방식이 비효율적이라 websocket에 컨트롤러를 직접 추가하는게 나을 것 같은데
아니라면 webSocket 모듈은 Websocket 통신에만 집중하고 API Gateway 역할은 모두 app-child에서 처리하는 게 아키텍쳐 일관성 측면에서 좋은 것 같기도 해서요..! 어떤 방식이 적합할까요?!!?

happyAndy0412 replied:
api 모듈에서 웹소켓 모듈의 api를 호출하는게 의미적으로도 안맞고 결합도도 증가시킬것같다는 차원에서 그렇게 생각했습니다 차라리 api만 child 모듈로 분리하고 싶으시면 클라이언트에서 openmarketapi 호출하고 반환값으로 웹소켓모듈에 요청하는게 나을 것 같습니다

waitAmhee replied:
네네!! websocket 모듈로 옮기라는 말씀 맞으시죠? 수정하겠습니다! 👍

happyAndy0412 commented on app-child/src/main/java/com/fintory/child/exceptionhandler/GlobalExceptionHandler.java:
```diff
@@ -66,12 +67,11 @@ public ResponseEntity<ExceptionResponse> handleHandlerMethodValidation(HandlerMe
 
     @ExceptionHandler(Exception.class)
-    public ResponseEntity<ExceptionResponse> handleUnhandledException(Exception e) {
+    public ResponseEntity<ExceptionResponse> handleUnhandledException(Exception e, HttpServletRequest request) {
```
request객체가 사용중이지 않은 것으로 보이는데 추후에 사용하실 계획이신가요?

waitAmhee replied:
아 이 때 프로메테우스 에러 로그가 서버 콘솔에 안 찍히게 하려고 수정했다가 돌리는 과정에서 제가 못 지운 것 같네요! 수정하겠습니다!! 👍

happyAndy0412 commented on app-child/src/main/resources/application-deploy.yml:
```diff
@@ -6,6 +6,10 @@ spring:
     activate:
       on-profile: deploy
 
+  websocket:
```
이부분 websocket 모듈 내의 application-deploy.yml에 있어야 하는 부분 아닌가요? config도 그쪽에 있던데 확인 부탁드립니다

waitAmhee replied:
앗! 제가 작성한 부분이 아니라서 검토할 때 놓쳤네요! 옮겨 놓도록 하겠습니다 !! 👍

happyAndy0412 commented on app-child/src/main/resources/application-local.yml:
```diff
@@ -100,6 +100,9 @@ db-openapi:
 firebase:
   config: ${FIREBASE_CONFIG}
 
+websocket:
```
이부분도 만약 CommonStockControllerImpl.getMarketStatus 부분 websocket모듈로 옮기게 되면 해당 파일에는 필요없을거같습니다

happyAndy0412 commented on domain/src/main/java/com/fintory/domain/stock/dto/websocket/LiveStockPriceStream.java:
```diff
@@ -2,6 +2,7 @@
+import java.io.Serializable;
 import java.math.BigDecimal;
 
 @JsonIgnoreProperties(ignoreUnknown = true)
```
저희 이거 혹시 현재가랑 코드만 가져오는 구조 아니었나요?

waitAmhee replied:
저는 프론트쪽에서 화면에 주식 변동률을 표시할 때 priceChangeRate 부분을 사용하고 있는 것으로 알고 있는데 아니라면 삭제하겠습니다!!

happyAndy0412 replied:
변동률 프론트에서 계산하는 방식 아니었나요?

waitAmhee replied:
웹소켓 모드에서는  증권사 서버에서 보내주는 값(realtimePriceChangeRate)이 있으면 그 값을 바로 사용하고 없으면 계산해서 화면에 보여주더라구요!  데이터 크기가 크지 않고, 정확도를 위해서는 현재 방식을 사용하는게 좋을 것 같은데 어떠세요?! 그리고 priceChange는 삭제하도록 하겠습니다! 👍 (혹시 제가 프론트 코드를 잘못 이해한거라면...죄송합니다 수정할게용 - ChangeRateStockListProps.tsx)

happyAndy0412 commented on infra/src/main/java/com/fintory/infra/domain/alarm/config/RedisSubscribeConfig.java:
```diff
@@ -0,0 +1,36 @@
+ */
+@Configuration
+@RequiredArgsConstructor
+public class RedisSubscribeConfig {
```
해당 config는 fcm 알림용인가요? 아니면 웹소켓 통신도 겸용인가요? alarm패키지에 있어서 궁금합니다

waitAmhee replied:
fcm 알림 트리거를 위한 redis pub/sub입니다!(ec2-websocket에서 현재가를 받으면 ec2-appchild에서 fcm 알림을 발송 )  Redis 채널에서 메시지가 publish되면 listener의 onMessage 함수가 트리거됨 -> 받은 메시지를 역직렬화해서 가격 알림 조건을 체크하고 알림을 보내는 방식입니다!

happyAndy0412 commented on websocket/build.gradle:
```diff
@@ -0,0 +1,80 @@
+}
+
+jar {
+    enabled = true //다른 모듈이 참조할 수 있는 일반 jar로 변경함
```
어떤 모듈이 참조하나요? bootjar는 참조가 불가능한가요?

waitAmhee replied:
bootJar로는 다른 모듈이 참조가 불가능한데, 
원래는 app-child에서 websocket 모듈을 참조하도록 예정>  코드가 수정되면서 해당 코드도 필요없게 되었어요! 
삭제하도록 하겠습니다! 👍

happyAndy0412 commented on websocket/src/main/java/com/fintory/websocket/config/SecurityConfig.java:
```diff
@@ -0,0 +1,67 @@
+
+import java.util.Arrays;
+import java.util.List;
+
```
웹소켓 관련 엔드포인트들 다 public으로 뚫어놨는데 시큐리티를 또 붙인 이유가 궁금합니다
웹소켓 성능 테스트의 결합도를 떨어뜨리기위해 보안을 적용하지 않는 방향성은 어떠신가요

waitAmhee replied:
Spring security는 프로메테우스 때문에 추가했습니다!
제가 처음에 실행했을 때 target 페이지에서 -> expected a valid start token, got "<" ("INVALID") 에러가 발생했어요! 이건 /actuator/prometheus 엔드포인트에 접근했을 때 html 페이지를 받았다는 뜻이라는데 spring의 기본 보안 정책 때문에 /actuator/prometheus가 보호가 되어서 인증 없이 접근했을 때의 에러라고 합니다! app-child 모듈용 spring security를 사용하기에는 관련 클래스들을 전부 스캔해야 하다보니최소한의 security 설정만 추가하려고(단순히 actuator를 노출) websocket용을 따로 만들었어요.

happyAndy0412 commented on websocket/src/main/java/com/fintory/websocket/provider/handler/KoreanLiveStockPriceWebSocketHandler.java:
```diff
@@ -68,14 +68,8 @@ public boolean waitForConnection(long timeoutSeconds){
     @Override
```
setSaveCallBack 메소드가 사용중이지 않는것으로 보이는데 확인 부탁드려요

ugiuk00 commented on websocket/src/main/java/com/fintory/websocket/provider/service/WebSocketConnectionService.java:
```diff
@@ -0,0 +1,203 @@
+@Slf4j
+public class WebSocketConnectionService {
+
+    private final RedisTemplate<Object, Object> redisTemplate;
```
사용중이지 않습니다

happyAndy0412 commented on websocket/src/main/java/com/fintory/websocket/publisher/state/StockDataHolder.java:
```diff
@@ -0,0 +1,35 @@
+package com.fintory.websocket.publisher.state;
+
```
해당 클래스 목적이 궁금합니다

waitAmhee replied:
하나의 클래스를 여러 클래스로 분리하는 과정 -> Websocket 관련 기능 클래스들이 공통으로 쓰는 객체들을 공동으로 관리하기 위한 클래스입니당

happyAndy0412 reviewed (approved):
리팩토링 수고했습니다 👍

waitAmhee committed:
fix : websocket stomp 설정 옮김

waitAmhee committed:
fix : app-child에서 websocket 모듈 참조 x

waitAmhee committed:
refactor: 사용하지 않는 코드 삭제

happyAndy0412 reviewed (commented):
👍 👍

waitAmhee committed:
refactor: 중복 stock 데이터 필터링 로직을 테스트를 위해 잠시 주석 처리

waitAmhee committed:
feat: .env 방식으로 deploy-websocket도 변경

Replaced individual export statements with a .env file creation for environment variables.
